File: MaximumProductSubarray.js
var maxProduct = function(nums) {
    let maxProduct = nums[0]
    let maxCurrProduct = nums[0]
    let minCurrProduct = nums[0]
    for(let i=1;i<nums.length;i++){
        let temp = Math.max(nums[i],maxCurrProduct*nums[i],minCurrProduct*nums[i])
        minCurrProduct = Math.min(nums[i],maxCurrProduct*nums[i],minCurrProduct*nums[i])
        maxProduct = Math.max(maxProduct,maxCurrProduct)
        maxCurrProduct = temp
        maxProduct = Math.max(maxCurrProduct,maxProduct)
    }
    return maxProduct
};

File: MaximumProductSubarray.py
class MaximumProductSubarray(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        maxPositive = nums[0]
        maxNegative = nums[0]
        maxProd = nums[0]
        for i in range(1, len(nums)):
            temp = max(nums[i], maxPositive * nums[i], maxNegative * nums[i])
            maxNegative = min(nums[i], maxPositive * nums[i], maxNegative * nums[i])
            maxProd = max(maxProd, maxPositive)
            maxPositive = temp
            maxProd = max(maxProd, maxPositive)
                
        return maxProd